### PR TITLE
Slightly improve quantization and add test

### PR DESCRIPTION
**Pull request by ibab**
The current quantization procedure is slightly broken, as it introduces a small bias when encoding and decoding the signal.
I've modified it so that this doesn't happen and added a test.

### CODE REVIEW

File: test/test_mu_law.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import tensorflow as tf
```
Comment (lemonzi) [inline test/test_mu_law.py]: Should be:

```python
import numpy as np
import tensorflow as tf

from wavenet import WaveNet
```

Unless we treat wavenet as a separate package, in which case the empty line should be removed.

File: test/test_mu_law.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import numpy as np
+from wavenet import WaveNet
+
+class TestMuLaw(tf.test.TestCase):
+
+    def testEncodeDecode(self):
+
+        net = WaveNet(
+            batch_size=1,
+            channels=256,
```
Comment (lemonzi) [inline test/test_mu_law.py]: `quantization_channels`, and as the last argument for clarity.

File: wavenet.py [hunk 3]
```diff
@@ -65,25 +65,29 @@ def _create_dilation_layer(self, input_batch, layer_index, dilation, in_channels
         return transformed, input_batch + transformed
 
 
-    def _preprocess(self, audio):
+    def encode(self, audio):
         '''Quantizes waveform amplitudes.'''
-        with tf.name_scope('preprocessing'):
+        with tf.name_scope('encode'):
             mu = self.channels - 1
```
Comment (lemonzi) [inline wavenet.py]: `self.quantization_steps` (will come with the upstream merge).

File: wavenet.py [hunk 4]
```diff
@@ -65,25 +65,29 @@ def _create_dilation_layer(self, input_batch, layer_index, dilation, in_channels
...
 
     def decode(self, output):
-        mu = self.channels - 1
-        y = tf.cast(output, tf.float32)
-        y = 2 * (y / mu) - 1
-        x = tf.sign(y) * (1 / mu) * ((1 + mu)**abs(y) - 1)
-        return x
+        '''Recovers waveform from quantized values.'''
+        with tf.name_scope('decode'):
+            mu = self.channels - 1
```
Comment (lemonzi) [inline wavenet.py]: `quantization_channels`

File: test/test_mu_law.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+class TestMuLaw(tf.test.TestCase):
+
+    def testEncodeDecode(self):
+
+        net = WaveNet(
+            batch_size=1,
+            dilations=[1],
+            filter_width=2,
+            quantization_channels=256,
```
Comment (lemonzi) [inline test/test_mu_law.py]: It should be the last argument

File: wavenet.py [hunk 6]
```diff
@@ -106,22 +106,24 @@ def _create_dilation_layer(self,
...
-        x = tf.sign(y) * (1 / mu) * ((1 + mu)**abs(y) - 1)
-        return x
+        '''Recovers waveform from quantized values.'''
+        with tf.name_scope('decode'):
+            mu = self.quantization_channels - 1
+            # Map values back to [-1, 1]
+            casted = tf.cast(output, tf.float32)
+            y = 2 * (casted / mu) - 1
+            # Perform inverse of mu-law transformation
+            return tf.sign(y) * (1 / mu) * ((1 + mu)**abs(y) - 1)
```
Comment (mecab) [inline wavenet.py]: Is it intentional no space? 
```
(1 + mu)**abs(y)
        ^^
```

Comment (ibab) [inline wavenet.py]: I like putting no space around the exponentiation as an exception to the rule.
This is just a habit from writing a lot of scientific code.
What are people's opinions on this?

Comment (lemonzi) [inline wavenet.py]: The line is also very long, maybe mimic `encode` with magnitude/signal?

Comment (lemonzi) [inline wavenet.py]: Since we may add it to Travis... does the PEP8 linter complain?

**Comment by lemonzi** [comment]
There should also be a test that checks that supplying an invalid `quanization_channels` (that is, <= 1) results in a predictable error, and possibly an assertion in the code that throws it.

File: wavenet.py (hunk 6 above)
Comment (mecab) [inline wavenet.py]: Yes, I understand putting no space around `**` is reasonable.
In terms of PEP8, it force putting space on some operators, but not for `**` (and I've just ran pep8) , so it's okay 👍

Let's go for it.

File: test/test_mu_law.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import numpy as np
+import tensorflow as tf
+
+from wavenet import WaveNet
+
```
Comment (mecab) [inline test/test_mu_law.py]: Needs 2 spaces before class.

**Comment by ibab** [comment]
How about `mu_law_encode` and `mu_law_decode`?

File: wavenet_ops.py [hunk 8]
```diff
@@ -35,3 +35,26 @@ def causal_conv(value, filter_, dilation, name='causal_conv'):
                          [0, 0, 0],
                          [-1, tf.shape(value)[1], -1])
         return result
+
```
Comment (mecab) [inline wavenet_ops.py]: Only one line break is allowed before function.

(Sorry for pointing out only small things... 🙇)

Comment (ibab) [inline wavenet_ops.py]: I'm not sure what you mean, could you explain?

Comment (mecab) [inline wavenet_ops.py]: Well, maybe the line I commented has been moved after the following commit.

File: wavenet_ops.py [hunk 9]
```diff
@@ -32,6 +34,29 @@ def causal_conv(value, filter_, dilation, name='causal_conv'):
             restored = tf.nn.conv1d(padded, filter_, stride=1, padding='SAME')
         # Remove excess elements at the end
         result = tf.slice(restored,
-                         [0, 0, 0],
-                         [-1, tf.shape(value)[1], -1])
+                          [0, 0, 0],
+                          [-1, tf.shape(value)[1], -1])
         return result
+
+
```
Comment (mecab) [inline wavenet_ops.py]: I was about to comment here.

> Only one line break is allowed before function.

Comment (mecab) [inline wavenet_ops.py]: Something is strange in Github.

I want to comment on `wavenet_ops.py` L42
```


def mu_law_encode(audio, quantization_channels):
```
should be 
```

def mu_law_encode(audio, quantization_channels):
```

File: test/test_mu_law.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
+import numpy as np
+import tensorflow as tf
+
+from wavenet_ops import mu_law_encode, mu_law_decode
+
+
+class TestMuLaw(tf.test.TestCase):
+
+    def testEncodeDecode(self):
```
Comment (lemonzi) [inline test/test_mu_law.py]: Could you add the `len(np.unique(x)) == quantization_steps` assertion?

Comment (ibab) [inline test/test_mu_law.py]: Is the code that I've added just now what you had in mind?

Comment (lemonzi) [inline test/test_mu_law.py]: Yep, this is perfect, thanks!

File: wavenet_ops.py (hunk 9 above)
Comment (ibab) [inline wavenet_ops.py]: Ah, I understand.
But shouldn't there be 2 lines there according to PEP8?

Comment (lemonzi) [inline wavenet_ops.py]: Yep, I think it's 1 line between methods or inner functions, but 2 lines for top-level functions and classes.

File: wavenet_ops.py [hunk 11]
```diff
@@ -32,6 +34,29 @@ def causal_conv(value, filter_, dilation, name='causal_conv'):
...
+def mu_law_encode(audio, quantization_channels):
+    '''Quantizes waveform amplitudes.'''
+    with tf.name_scope('encode'):
+        mu = quantization_channels - 1
+        # Perform mu-law companding transformation (ITU-T, 1988).
+        magnitude = tf.log(1 + mu * tf.abs(audio)) / tf.log(1. + mu)
+        signal = tf.sign(audio) * magnitude
+        # Quantize signal to the specified number of levels
```
Comment (lemonzi) [inline wavenet_ops.py]: Add periods at the end of the comments?

Comment (lemonzi) [inline wavenet_ops.py]: Besides that (which we can ignore if you feel lazy about it), it LGTM. There's some docstrings missing, but we can add them later.

**Comment by jyegerlehner** [comment]
LGTM. Yay, tests.

**Pull request merged**